### ISSUE TITLE
Proposal: Review of goroutine/channel communications

### ISSUE THREAD

**Issue by MattBrittan** [eclipse/paho.mqtt.golang#372]
I'd like to propose a review of the way the various goroutines used to co-ordinate incomming/outgoing traffic work together. I believe that there are a few issues with the way this currently works:

* leads to a number of potential deadlock (e.g. the one mentioned in #369 and a few with c.errors)
* makes modifications fairly complex (because it's easy to miss the need to check for a closed stop channel when sending anything to obound/oboundP etc).
* complicates debugging (particularly tracing deadlocks)

I believe that this could be simplified, and potential deadlocks removed, by rewriting the logic such that the various go routines are independent (not accessing the client class directly) and shutdown in a more managed fashion (i.e. each goroutine ranges a channel and only terminates when that channel has been closed).

I'm prepared to make the changes but as it would be a fairly major change (with the potential to create new issues!) I thought it would be best to discuss it prior to proceeding.

## Summary of current goroutines

Incomming - receives from c.stop. Sends to ibound and errors
Outgoing - receives from obound, oboundP and c.stop. Sends to errors
Alllogic - receives from ibound and c.stop. Sends to obound, oboundP and incommingPubChan
matchAndDispatch - receives from incommingPubChan and r.stop (different from c.stop). Sends to oboundP (via c.ackFunc)
errorWatch -  receives from errors and c.stop (via internalConnLost).

These routines are syncronised using a waitgroup (other than matchAndDispatch which survives internal disconnections).

**Comment by alsm** [comment]
I'm happy for you to take a look at doing this work, no doubt there are simplifications that can be made to improve the quality and the performance of the client.

**Comment by MattBrittan** [comment]
Thanks @alsm,
I have gone ahead and raised pull request #381 ; however recognise that this is a major change (somewhat larger I expected!) and others may not like my preferred way of dealing with goroutine syncronisation (or possibly just think my code is rubbish!). On that basis I've submitted this in a state where its passing the existing tests (with small modification due to a change in the way the router is run); if you are amenable to the changes I'll review everything again and test further.

Many of the changes I've made are really tidying up areas that have been patched multiple times over the years. My aim has been to reduce the areas where deadlocks can occur so that future changes will be less likely to cause them (but am sure I've added a few of my own) and add comments to aid future debugging. I had looked at trying to make the code more similar to your v5 client but the need to maintain the existing interface makes this very difficult.

I'm will not be concerned if this gets rejected as it's a big change to make to a mature project that is already mostly reliable (I'm sure I will have introduced bugs). It's been a good learning experience for me and I certainly understand the existing codebase better that I did.
